core/src/main/java/com/linecorp/armeria/server/CorsServerErrorHandler.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2024 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License
+ */
+
+package com.linecorp.armeria.server;
+
+import com.linecorp.armeria.common.AggregatedHttpResponse;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.common.ResponseHeaders;
+import com.linecorp.armeria.common.ResponseHeadersBuilder;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.server.CorsHeaderUtil;
+import com.linecorp.armeria.server.cors.CorsConfig;
+import com.linecorp.armeria.server.cors.CorsService;
+
+/**
+ * wraps ServerErrorHandler for adding CORS headers to error responses.
+ */
+public class CorsServerErrorHandler implements ServerErrorHandler {
+    ServerErrorHandler serverErrorHandler;
+
+    /**
+     * Constructs a new {@link CorsServerErrorHandler} instance with a specified {@link ServerErrorHandler}.
+     * This handler is used to delegate server error handling for CORS-related errors.
+     *
+     * @param serverErrorHandler The {@link ServerErrorHandler} to be used for handling server errors.
+     */
+    public CorsServerErrorHandler(ServerErrorHandler serverErrorHandler) {
+        this.serverErrorHandler = serverErrorHandler;
+    }
+
+    @Override
+    public @Nullable AggregatedHttpResponse renderStatus(@Nullable ServiceRequestContext ctx,
+                                                         ServiceConfig serviceConfig,
+                                                         @Nullable RequestHeaders headers,
+                                                         HttpStatus status, @Nullable String description,
+                                                         @Nullable Throwable cause) {
+
+        final CorsService corsService = serviceConfig.service().as(CorsService.class);
```

Comment:
We can't simply find a `CorsService` added by `ServerBuilder.decorate()` with `.as()` operation.

All decorators would be searched from `Router` instead of using `.as()`. A possible workaround is to add a helper method to `InitialDispatcherService`.
```java
public static class InitialDispatcherService extends SimpleDecoratingHttpService {

  @Nullable
  public <T extends HttpService> T findService(ServiceRequestContext ctx, Class<? extends T> serviceClass) {
    final List<Routed<RouteDecoratingService>> serviceChain = router.findAll(ctx.routingContext());
    if (serviceChain.isEmpty()) {
        return unwrap().as(serviceClass);
    }
    
    for (Routed<RouteDecoratingService> routed : serviceChain) {
        if (routed.isPresent()) {
            final HttpService service = routed.value().decorator();
            final T targetService = service.as(serviceClass);
            if (targetService != null) {
                return targetService;
            }
        }
    }
    return null;
}
}
```

Reply:
Created https://github.com/line/armeria/pull/5670 
If the PR is merged, we can the code with:
```java
final CorsService corsService = ctx.findService(CorsService.class);
```

Reply:
Let me add a follow-up commit once it's merged. :smile: